packages/ckeditor5-source-editing/src/sourceediting.js

```diff
@@ -335,3 +335,105 @@ export default class SourceEditing extends Plugin {
 		return editable && !editable._hasExternalElement;
 	}
 }
+
+// Beautifies the input HTML string by adding new lines and indenting some HTML elements.
+//
+// @param {String} input HTML string to beautify.
+// @returns {String}
+function formatSource( input ) {
+	const elementsToFormat = [
+		{ name: 'blockquote', isVoid: false },
+		{ name: 'br', isVoid: true },
+		{ name: 'div', isVoid: false },
+		{ name: 'figcaption', isVoid: false },
+		{ name: 'figure', isVoid: false },
+		{ name: 'h1', isVoid: false },
+		{ name: 'h2', isVoid: false },
+		{ name: 'h3', isVoid: false },
+		{ name: 'h4', isVoid: false },
+		{ name: 'h5', isVoid: false },
+		{ name: 'h6', isVoid: false },
+		{ name: 'hr', isVoid: true },
+		{ name: 'li', isVoid: false },
+		{ name: 'ol', isVoid: false },
+		{ name: 'p', isVoid: false },
+		{ name: 'table', isVoid: false },
+		{ name: 'tbody', isVoid: false },
+		{ name: 'td', isVoid: false },
+		{ name: 'th', isVoid: false },
+		{ name: 'thead', isVoid: false },
+		{ name: 'tr', isVoid: false },
+		{ name: 'ul', isVoid: false }
+	];
+
+	const elementNamesToFormat = elementsToFormat.map( element => element.name ).join( '|' );
+
+	const lines = input
+		// Add new line before `<tag>` or `</tag>`, but only if it is not already preceded by a new line (negative lookbehind).
+		.replace( new RegExp( `(?<!\n)</?(${ elementNamesToFormat })( .*?)?>`, 'g' ), '\n$&' )
+		// Add new line after `<tag>` or `</tag>`, but only if it is not already followed by a new line (negative lookahead).
+		.replace( new RegExp( `</?(${ elementNamesToFormat })( .*?)?>(?!\n)`, 'g' ), '$&\n' )
+		// Divide input string into lines, which start with either an opening tag, a closing tag, or just a text.
+		.split( '\n' );
```

Comment:
I think that this may be slow for large documents with the `elementNamesToFormat` being an alternation of many strings. But I might be wrong.

Reply:
I've tested the performance of `formatSource()` function with generated random HTML content (paragraphs, tables, lists, etc.):

| HTML content size | Execution time |
| --- | --- |
| 100 KB | ~100 ms |
| 200 KB | ~190 ms |
| 500 KB | ~460 ms |
| 1 MB | ~900 ms |

IMHO these results are good enough.

Reply:
Currently, it takes about half of the time of `editor.data.get()`:

```
Editor.getData() 109
formatSource() 55

Editor.getData() 85
formatSource() 55
```

Reply:
I'm not sure if it's OK or NOK but it can be easily improved later if this would be a problem.